File: haggis-platform/client/Haggis.Flutter.UI/lib/local_game/ai/heuristic_weight_normalization.dart
import 'dart:math' as math;

/// Exact counterpart of HeuristicWeightNormalization in Haggis.AI.
abstract final class HeuristicWeightNormalization {
  static const int maxHandSize = 17;

  static int baseScore(double normalizedSignal, int baseImpact) {
    if (baseImpact <= 0) return 0;
    return _roundAwayFromZero(clamp(normalizedSignal, -1, 1) * baseImpact);
  }

  static int applyWeight(int baseScore, double weight) {
    if (weight <= 0 || baseScore == 0) return 0;
    return _roundAwayFromZero(baseScore * weight);
  }

  static double normalize(double value, double min, double max) {
    if (max <= min) return 0;
    return clamp((value - min) / (max - min), 0, 1);
  }

  static double normalizeRatio(num numerator, num denominator) {
    if (denominator <= 0) return 0;
    return clamp(numerator / denominator, 0, 1);
  }

  static double handPhase(int handCount) =>
      normalizeRatio(handCount, maxHandSize);

  static double clamp(double value, double min, double max) =>
      math.max(min, math.min(max, value));

  static int _roundAwayFromZero(double value) =>
      value >= 0 ? (value + 0.5).floor() : (value - 0.5).ceil();
}

File: haggis-platform/backend/src/Haggis.AI/HeuristicEngine/WeightNormalization/HeuristicWeightNormalization.cs
using System;

namespace Haggis.AI.WeightNormalization
{
    internal static class HeuristicWeightNormalization
    {
        public const int MaxHandSize = 17;

        public static int BaseScore(double normalizedSignal, int baseImpact)
        {
            if (baseImpact <= 0)
            {
                return 0;
            }

            var clampedSignal = Clamp(normalizedSignal, -1d, 1d);
            return (int)Math.Round(clampedSignal * baseImpact, MidpointRounding.AwayFromZero);
        }

        public static int ApplyWeight(int baseScore, float weight)
        {
            if (weight <= 0f || baseScore == 0)
            {
                return 0;
            }

            return (int)Math.Round(baseScore * weight, MidpointRounding.AwayFromZero);
        }

        public static double Normalize(double value, double min, double max)
        {
            if (max <= min)
            {
                return 0d;
            }

            return Clamp((value - min) / (max - min), 0d, 1d);
        }

        public static double NormalizeRatio(double numerator, double denominator)
        {
            if (denominator <= 0d)
            {
                return 0d;
            }

            return Clamp(numerator / denominator, 0d, 1d);
        }

        public static double HandPhase(int handCount)
        {
            return NormalizeRatio(handCount, MaxHandSize);
        }

        public static double Clamp(double value, double min, double max)
        {
            return Math.Max(min, Math.Min(max, value));
        }
    }
}
